src/Swift.Runtime/src/Swift/SwiftString.cs

```diff
@@ -0,0 +1,155 @@
+    }
+
+    static ISwiftObject ISwiftObject.NewFromPayload(SwiftHandle handle)
+    {
+        return new SwiftString(handle);
+    }
+
+    IntPtr ISwiftObject.MarshalToSwift(IntPtr swiftDest)
+    {
+        var metadata = SwiftObjectHelper<SwiftString>.GetTypeMetadata();
+        unsafe
+        {
+            fixed (void* _payloadPtr = &_payload)
+            {
+                metadata.ValueWitnessTable->InitializeWithCopy((void*)swiftDest, (void*)_payloadPtr, metadata);
+            }
+        }
+        return swiftDest;
+    }
+
+    /// <summary>
+    /// Gets the protocol conformance descriptor for the given type.
+    /// </summary>
+    /// <typeparam name="TProtocol"></typeparam>
+    /// <returns></returns>
+    static ProtocolConformanceDescriptor ISwiftObject.GetProtocolConformanceDescriptor<TProtocol>()
+        where TProtocol : class
+    {
+        if (!_protocolConformanceSymbols.TryGetValue(typeof(TProtocol), out var symbolName))
+        {
+            throw new SwiftRuntimeException($"Attempted to retrieve protocol conformance descriptor for type SwiftString and protocol {typeof(TProtocol).Name}, but no conformance was found.");
+        }
+        return ProtocolConformanceDescriptor.LoadFromSymbol("/usr/lib/swift/libswiftCore.dylib", symbolName);
+    }
+
+    /// <summary>
+    /// Constructs a new SwiftString from the given handle.
+    /// </summary>
+    unsafe SwiftString(SwiftHandle handle)
+    {
+        _payload = *(Data*)handle;
+    }
+
+    /// <summary>
+    /// Constructs a new SwiftString from the C# string.
+    /// </summary>
+    public SwiftString(string str)
+    {
+        byte[] utf8Bytes = Encoding.UTF8.GetBytes(str);
+        unsafe
+        {
+            fixed (byte* utf8BytesPtr = utf8Bytes)
+            {
+                _payload = PInvoke_Create(utf8BytesPtr, utf8Bytes.Length, 1);
+            }
+        }
+    }
+
+    /// <summary>
+    /// Gets the length of string.
+    /// </summary>
+    public int Length => (int)PInvoke_GetLength(_payload);
+
+    /// <summary>
+    /// Converts the SwiftString to a C# string.
+    /// </summary>
+    public override string ToString()
+    {
+        unsafe
+        {
+            var elementType = TypeMetadata.GetTypeMetadataOrThrow<byte>();
+            var resultType = TypeMetadata.GetTypeMetadataOrThrow<long>();
+
+            var len = Length;
+            if (len <= 0)
+                return string.Empty;
+
+            string? str = null;
+
+            var arr = PInvoke_GetUtf8ContiguousArray(_payload);
+            PInvoke_WithUnsafeBytes(bytes =>
+                {
+                    unsafe
+                    {
+                        str = Encoding.UTF8.GetString((byte*)bytes, len);
+                        return IntPtr.Zero;
+                    }
+                }, IntPtr.Zero, arr, elementType, resultType);
+
+            return str!;
+        }
+    }
+
+    public unsafe delegate IntPtr CallbackDelegate(IntPtr param);
```

Comment:
Should the callback have default calling convention or Swift calling convention? (Related to my other comment about function pointers.)

Reply:
Also, should the callback signature have the closure `context` as the argument?

Reply:
Yes.

`ToString` has been refactored to use function pointers. The `length` is passed through the closure context to the callback via the self register. A utf8 buffer is allocated, and the content of the contiguous array is copied into it. The buffer is then marshalled using `PtrToStringUTF8`, and the native buffer is deallocated.